File: AlwaysPrintProject/Cloud/backend/app/schemas/knowledge_article.py
"""
Schemas Pydantic para artículos de conocimiento.

Estos schemas validan la entrada/salida de la API de artículos técnicos
que se inyectan como contexto adicional en el prompt del LLM durante
el análisis de sesiones de debugging.
"""

from datetime import datetime
from typing import List, Optional
from uuid import UUID
from pydantic import BaseModel, Field, field_validator


# === SCHEMAS DE REQUEST ===


class KnowledgeArticleCreate(BaseModel):
    """Schema para crear un artículo de conocimiento."""

    title: str = Field(
        ..., min_length=3, max_length=200, description="Título del artículo"
    )
    description: str = Field(
        ..., min_length=10, max_length=500, description="Descripción breve"
    )
    content: str = Field(
        ..., min_length=1, max_length=500_000, description="Contenido Markdown"
    )

    @field_validator("content")
    @classmethod
    def validate_content_not_empty(cls, v: str) -> str:
        """Validar que el contenido no sea solo espacios en blanco."""
        if not v.strip():
            raise ValueError("El contenido no puede estar vacío")
        return v


class KnowledgeArticleUpdate(BaseModel):
    """Schema para actualizar un artículo existente (campos opcionales)."""

    title: Optional[str] = Field(None, min_length=3, max_length=200)
    description: Optional[str] = Field(None, min_length=10, max_length=500)
    content: Optional[str] = Field(None, min_length=1, max_length=500_000)

    @field_validator("content")
    @classmethod
    def validate_content_not_empty(cls, v: Optional[str]) -> Optional[str]:
        """Validar que el contenido no sea solo espacios en blanco (si se proporciona)."""
        if v is not None and not v.strip():
            raise ValueError("El contenido no puede estar vacío")
        return v


# === SCHEMAS DE RESPONSE ===


class KnowledgeArticleResponse(BaseModel):
    """Schema de respuesta completa de un artículo."""

    id: UUID
    organization_id: UUID
    title: str
    description: str
    content: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class KnowledgeArticleListItem(BaseModel):
    """Schema resumido para listados (sin contenido completo)."""

    id: UUID
    title: str
    description: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# === SCHEMAS DE ASOCIACIÓN ===


class ProfileArticleAssociation(BaseModel):
    """Schema para asociar artículos a un perfil de debugging."""

    article_ids: List[UUID] = Field(
        ...,
        min_length=1,
        max_length=10,
        description="Lista de IDs de artículos a asociar (máximo 10)",
    )

File: AlwaysPrintProject/Cloud/frontend/src/types/knowledge-article.ts
/**
 * Tipos relacionados con artículos de conocimiento (Knowledge Base Library).
 *
 * Los artículos almacenan documentación técnica en Markdown que se inyecta
 * como contexto adicional en el prompt del LLM durante el análisis de debugging.
 */

/**
 * Artículo de conocimiento completo (respuesta del backend).
 */
export interface KnowledgeArticle {
  /** Identificador único del artículo (UUID) */
  id: string
  /** Identificador de la organización propietaria (UUID) */
  organization_id: string
  /** Título del artículo (3-200 caracteres) */
  title: string
  /** Descripción breve del artículo (10-500 caracteres) */
  description: string
  /** Contenido en formato Markdown (máx 500KB) */
  content: string
  /** Fecha de creación (ISO 8601) */
  created_at: string
  /** Fecha de última actualización (ISO 8601) */
  updated_at: string
}

/**
 * Datos requeridos para crear un artículo de conocimiento.
 */
export interface KnowledgeArticleCreate {
  /** Título del artículo (3-200 caracteres) */
  title: string
  /** Descripción breve del artículo (10-500 caracteres) */
  description: string
  /** Contenido en formato Markdown (máx 500KB) */
  content: string
}

/**
 * Datos opcionales para actualizar un artículo existente.
 * Solo los campos proporcionados se actualizan.
 */
export interface KnowledgeArticleUpdate {
  /** Título del artículo (3-200 caracteres) */
  title?: string
  /** Descripción breve del artículo (10-500 caracteres) */
  description?: string
  /** Contenido en formato Markdown (máx 500KB) */
  content?: string
}

/**
 * Versión resumida de un artículo para listados (sin contenido completo).
 */
export interface KnowledgeArticleListItem {
  /** Identificador único del artículo (UUID) */
  id: string
  /** Título del artículo */
  title: string
  /** Descripción breve del artículo */
  description: string
  /** Fecha de creación (ISO 8601) */
  created_at: string
  /** Fecha de última actualización (ISO 8601) */
  updated_at: string
}

/**
 * Request para asociar artículos de conocimiento a un perfil de debugging.
 */
export interface ProfileArticleAssociation {
  /** Lista de IDs de artículos a asociar (UUID[], máximo 10) */
  article_ids: string[]
}
